diff --git a/lib/index.cjs b/lib/index.cjs
index c9491f7dc24fc1f00033719e28ab9a5b73b4f96b..6a37f1a0b5e1bcd9f8b878901b67680e81143a57 100644
--- a/lib/index.cjs
+++ b/lib/index.cjs
@@ -33,7 +33,8 @@ const TableInput = (props) => {
         fontSize: 1,
         padding: 3,
         value: cell,
-        onChange: (e) => updateCell(e, rowIndex, cellIndex)
+        onChange: (e) => updateCell(e, rowIndex, cellIndex),
+        onKeyDown: (e) => e.stopPropagation()
       }
     ) }, `cell-${rowIndex}-${cellIndex}`);
   }, renderRow = (row, rowIndex) => {
@@ -261,8 +262,7 @@ const deepClone = (_a = globalThis.structuredClone) != null ? _a : (data) => JSO
   }, updateCell = (e, rowIndex, cellIndex) => {
     if (!value)
       return;
-    const newValue = deepClone(value);
-    return newValue.rows[rowIndex].cells[cellIndex] = e.target.value, updateValue(newValue);
+    return onChange(sanity.set(e.target.value, ["rows", { _key: value.rows[rowIndex]._key }, "cells", cellIndex]));
   };
   return /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
     dialog && /* @__PURE__ */ jsxRuntime.jsx(
diff --git a/lib/index.js b/lib/index.js
index 5d8dfb1306227c8ff9858bde90456a79a49635b8..d5f49799d0a5c75e18cf5b877fb5f2daee614f07 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -36,7 +36,8 @@ const TableInput = (props) => {
         fontSize: 1,
         padding: 3,
         value: cell,
-        onChange: (e) => updateCell(e, rowIndex, cellIndex)
+        onChange: (e) => updateCell(e, rowIndex, cellIndex),
+        onKeyDown: (e) => e.stopPropagation()
       }
     ) }, `cell-${rowIndex}-${cellIndex}`);
   }, renderRow = (row, rowIndex) => {
@@ -264,8 +265,7 @@ const deepClone = (_a = globalThis.structuredClone) != null ? _a : (data) => JSO
   }, updateCell = (e, rowIndex, cellIndex) => {
     if (!value)
       return;
-    const newValue = deepClone(value);
-    return newValue.rows[rowIndex].cells[cellIndex] = e.target.value, updateValue(newValue);
+    return onChange(set(e.target.value, ["rows", { _key: value.rows[rowIndex]._key }, "cells", cellIndex]));
   };
   return /* @__PURE__ */ jsxs("div", { children: [
     dialog && /* @__PURE__ */ jsx(
